discovery/wh_discovery.go

```diff
@@ -3,80 +3,148 @@ package discovery
 import (
 	"encoding/json"
 	"errors"
+	"fmt"
 	"io/ioutil"
+	"math/rand"
 	"net/http"
 	"net/url"
 	"time"
 
+	"github.com/ethereum/go-ethereum/common"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/livepeer/go-livepeer/core"
+	lpTypes "github.com/livepeer/go-livepeer/eth/types"
 	"github.com/livepeer/go-livepeer/net"
 
 	"github.com/golang/glog"
 )
 
+var whRefreshInterval = 1 * time.Minute
+
 type webhookResponse struct {
 	Address string
 }
 
 type webhookPool struct {
-	node     *core.LivepeerNode
-	callback *url.URL
+	db                   *DBOrchestratorPoolCache
+	callback             *url.URL
+	hashPreviousResponse ethcommon.Hash
+	lastRequested        time.Time
 }
 
 func NewWebhookPool(node *core.LivepeerNode, callback *url.URL) *webhookPool {
-	return &webhookPool{node: node, callback: callback}
+	addresses, hash, err := cacheWebhookTranscoders(node, callback, ethcommon.BytesToHash([]byte{}))
+
+	if err != nil || len(addresses) == 0 {
+		glog.Error("Error getting urls from webhook: ", err)
+		return nil
+	}
+
+	return &webhookPool{
+		db:                   NewDBOrchestratorPoolCache(node),
+		callback:             callback,
+		lastRequested:        time.Now(),
+		hashPreviousResponse: hash,
+	}
+}
+
+func (w *webhookPool) GetURLs() []*url.URL {
+	return w.db.GetURLs()
+}
+
+func (w *webhookPool) Size() int {
+	return len(w.db.GetURLs())
+}
+
+func (w *webhookPool) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
+	var addrs []*url.URL
+
+	if time.Since(w.lastRequested) < whRefreshInterval {
+		// retrive addrs from DB if time since lastRequested is less than the refresh interval
+		glog.Error("IN TIME SINCE")
+		return w.db.GetOrchestrators(numOrchestrators)
+	}
+
+	// retrive addrs from webhook if time since lastRequested is more than the refresh interval
+	addresses, hash, err := cacheWebhookTranscoders(w.db.node, w.callback, w.hashPreviousResponse)
```

Comment:
The division of responsibilities between `cacheWebhookTranscoders` and `GetOrchestrators` is a little unclear. One way to improve the flow is to simplify `cacheWebhookTranscoders` so it only does the actual caching, eg the DB insertion, and extract out the actual webhook call into the body here. Following up on the earlier suggestion of a `getURLs` helper method, something like:

```golang
func (w *webhookPool) getURLs() ([]*url.URL, error) {
  w.mu.RLock() // see explanation below
  lastReq := w.lastRequested
  w.mu.RUnlock()
  if time.Since(lastReq) < whRefreshInterval {
    return w.db.getURLs()
  }
  addrs, hash, err := getURLs(w.callback)
  if err != nil {
    return w.db.getURLs()
  }
  // Since these instance vars could be modified by different threads
  // they probably need to be locked as well. Remember to to read-lock
  // accesses as necessary.
  w.mu.Lock()
  if hash != w.hashPreviousResponse {
    cacheAddressesInDB(w.db, addrs)
  }
  w.lastRequested = time.Now() // this needs to be set as well
  w.hashPreviousResponse = hash
  w.mu.Unlock()
  return addrs, nil
}
```

(This example isn't as tight as it could be, especially because of the lack of mutual exclusion around the webhook call and the blocking DB response. If you want to see a more robust method of implementing this, look through git history for the code controlling access around selectOrchestrator, prior to us adding the broadcast session manager.)

Reply:
Oh I see what you mean -  `refreshSession` is a good guide. Again, might be something for when we do do caching.

Reply:
We still need caching. We can just live without the persistence that the DB offers, since that only helps us at the very beginning, during start-up time if the webhook endpoint happens to be down. For other scenarios, an in-memory cache of orchestrator URLs is fine.